### PR TITLE
Refactor findpath

### PR DESCRIPTION
**Pull request by nevermore3**
## What type of PR is this?
- [ ] bug
- [ ] feature
- [x] enhancement

#### Description:

1、reduce AllPath & multiPairShortestPath 's memory usage
2、Integrate the function of conjunct path into the path operator to reduce data copying

for specific details, please refer to the comment in the file

image

image

The blue line is optimized
The green line is not optimized

## Checklist:
Tests:
- [x] Unit test(positive and negative cases)
- [ ] Function test
- [x] Performance test
- [ ] N/A

Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

## Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

### CODE REVIEW

**Comment by nevermore3** [comment]
> ## What type of PR is this?
> * [ ]  bug
> * [ ]  feature
> * [x]  enhancement
> 
> ## What problem(s) does this PR solve?
> #### Issue(s) number:
> #### Description:
> 1、reduce AllPath & multiPairShortestPath 's memory usage 2、Integrate the function of conjunct path into the path operator to reduce data copying
> 
> for specific details, please refer to the comment in the file
> 
> ## How do you solve it?
> ## Special notes for your reviewer, ex. impact of this fix, design document, etc:
> ## Checklist:
> Tests:
> 
> * [x]  Unit test(positive and negative cases)
> * [ ]  Function test
> * [x]  Performance test
> * [ ]  N/A
> 
> Affects:
> 
> * [ ]  Documentation affected (Please add the label if documentation needs to be modified.)
> * [ ]  Incompatibility (If it breaks the compatibility, please describe it and add the label.）
> * [ ]  If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
> * [ ]  Performance impacted: Consumes more CPU/Memory
> 
> ## Release notes:
> Please confirm whether to be reflected in release notes and how to describe:
> 
> > ex. Fixed the bug .....

image

File: src/graph/executor/algo/BFSShortestPathExecutor.cpp [hunk 1]
```diff
@@ -1,52 +1,214 @@
...
+  if (meetVids.empty()) {
+    return Status::OK();
+  }
+  size_t totalSize = meetVids.size();
+  size_t batchSize = totalSize / static_cast<size_t>(FLAGS_num_operator_threads);
+  std::vector<Value> batchVids;
+  batchVids.reserve(batchSize);
+  std::vector<folly::Future<DataSet>> futures;
+  for (size_t i = 0; i < totalSize; ++i) {
+    batchVids.push_back(meetVids[i]);
```
Comment (yixinglu) [inline src/graph/executor/algo/BFSShortestPathExecutor.cpp]: If you iterate the meetVids vector from end to start, you will need not to copy the contents to a new vector.

File: src/graph/executor/algo/BFSShortestPathExecutor.cpp [hunk 2]
```diff
@@ -1,52 +1,214 @@
...
+      Row row;
+      row.emplace_back(std::move(result));
+      ds.rows.emplace_back(std::move(row));
+    }
+  }
+  return ds;
+}
+
+std::unordered_multimap<Value, Path> BFSShortestPathExecutor::createPath(
+    std::vector<Value> meetVids, bool reverse, bool oddStep) {
```
Comment (yixinglu) [inline src/graph/executor/algo/BFSShortestPathExecutor.cpp:174]: need not to copy the vector of values.

File: src/graph/executor/algo/BFSShortestPathExecutor.cpp [hunk 3]
```diff
@@ -1,52 +1,214 @@
...
 
+  return folly::collect(futures).via(runner()).thenValue([this](auto&& resps) {
+    for (auto& resp : resps) {
+      currentDs_.append(std::move(resp));
+    }
+    return Status::OK();
+  });
+}
+
+DataSet BFSShortestPathExecutor::doConjunct(const std::vector<Value> meetVids, bool oddStep) {
```
Comment (yixinglu) [inline src/graph/executor/algo/BFSShortestPathExecutor.cpp:155]: `const std::vector<Value> &`

File: src/graph/executor/algo/MultiShortestPathExecutor.cpp [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+  pathNode_ = asNode<MultiShortestPath>(node());
+  terminationVar_ = pathNode_->terminationVar();
+
+  if (step_ == 1) {
+    init();
+  }
+
+  std::vector<folly::Future<Status>> futures;
+  auto leftFuture = folly::via(runner(), [this]() { return buildPath(false); });
+  auto rightFuture = folly::via(runner(), [this]() { return buildPath(true); });
```
Comment (yixinglu) [inline src/graph/executor/algo/MultiShortestPathExecutor.cpp:21]: I want to remind you that the `buildPath` will be executed in different threads carefully, don't share the same variables inside the function, especially the fields of the class.

File: src/graph/executor/algo/MultiShortestPathExecutor.cpp [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+  }
+  // set nextVid
+  const auto& nextVidVar = reverse ? pathNode_->rightVidVar() : pathNode_->leftVidVar();
+  setNextStepVid(currentPaths, nextVidVar);
+  return Status::OK();
+}
+
+DataSet MultiShortestPathExecutor::doConjunct(Interims::iterator startIter,
+                                              Interims::iterator endIter,
+                                              bool oddStep) {
```
Comment (yixinglu) [inline src/graph/executor/algo/MultiShortestPathExecutor.cpp:129]: it's better to let it be a `const` function for thread-safe usage.

File: src/graph/executor/algo/BFSShortestPathExecutor.cpp (hunk 2 above)
Comment (nevermore3) [reply src/graph/executor/algo/BFSShortestPathExecutor.cpp:174]: done

File: src/graph/executor/algo/BFSShortestPathExecutor.cpp (hunk 3 above)
Comment (nevermore3) [reply src/graph/executor/algo/BFSShortestPathExecutor.cpp:155]: done

File: src/graph/executor/algo/MultiShortestPathExecutor.cpp (hunk 4 above)
Comment (nevermore3) [reply src/graph/executor/algo/MultiShortestPathExecutor.cpp:21]: Agree, the previous version used local variables but may cause data copying, so use the  member variables

**Review by yixinglu — APPROVED** [review]

**Review by CPWstatic — DISMISSED** [review]

**Review by yixinglu — APPROVED** [review]

**Review by CPWstatic — APPROVED** [review]

**Pull request merged**